### ISSUE TITLE
Make the navigation block content available when JavaScript is disabled.

### ISSUE THREAD

**Issue by carolinan** [WordPress/gutenberg#69363]
## Description

I am copying this from a Trac ticket that was opened about the menu in Twenty Twenty-Four:
https://core.trac.wordpress.org/ticket/63031

The reporter's clients needs to browse with JavaScript disabled, but the navigation blocks mobile menu does not work without JavaScript.

I understand that the editors require JavaScript and that the workaround for that is to install the classic editor plugin.
I also recommended the user to change to any theme that is specifically built to work without JavaScript.

But perhaps the navigation block can still be improved to have a noscript fallback.

## Expected behaviour

I expect the menu content to be available with reduced functionality.

## Actual behaviour

The menu does not open.

### Step-by-step reproduction instructions

Place a navigation block with the mobile menu option enabled. Save.
Go to the front of the site.
Disable JavaScript -How you do this will depend on the browser.
Try to open the menu.

### Please confirm that you have searched existing issues in the repo.

- [x] Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

- [x] Yes

### Please confirm which theme type you used for testing.

- [x] Block
- [ ] Classic
- [ ] Hybrid (e.g. classic with theme.json)
- [ ] Not sure

**Comment by maltfield** [comment]
Can someone please indicate which file in this repo has the "navigation block" code?

**Comment by maltfield** [comment]
> Note: For additional context, I'm going to post the [original ticket contents](https://core.trac.wordpress.org/ticket/63031) below

 ### Description

The Twenty Twenty-Four hamburger menu (for site navigation on narrow/mobile displays) does not open when clicked/tapped on browsers with javascript disabled.

Note: I'm creating this ticket here on trac because the GitHub repo said to do so here https://github.com/WordPress/twentytwentyfour/blob/6d45adf1f4ecb688eb79dacb0f4f9eb34d9d6594/README.md#archived-repository

### Steps to reproduce

1. Download, Install, Open Tor Browser https://www.torproject.org/download/
1. Enable "Safest" mode https://tb-manual.torproject.org/security-settings/
1. Navigate to some Twenty Twenty-Four website (eg https://wp-themes.com/twentytwentyfour/)
1. Reduce the width of the browser until the hamburger menu in the top-right appears
1. Menu icon doesn't appear
1. Clicking the empty area where the menu *should* be does nothing 

#### Actual behavior

When javascript is disabled, clicking the hamburger menu does nothing.

#### Expected behavior

The hamburger menu should always open, even on security-hardened web browsers.

#### Why?

There are countless [at-risk](https://community.torproject.org/user-research/personas/) users around the world who must use Tor Browser for all Internet activity to protect themselves and/or their families. Examples of such groups include:

1. Refugees (eg escaping gang, State, or domestic violence)
1. Journalists (eg reporting on gangs, corruption, or oppressive regimes)
1. Whistleblowers (eg providing information to journalists above)
1. Activists (eg pro-abortion-rights or environmental activists) 

Unfortunately, every year numerous people are assassinated in the the above at-risk groups <sup>[[1](https://cpj.org/data/killed/2023/?status=Killed&motiveConfirmed%5B%5D=Confirmed&motiveUnconfirmed%5B%5D=Unconfirmed&type%5B%5D=Journalist&type%5B%5D=Media%20Worker&start_year=2023&end_year=2023&group_by=location)][[2](https://rsf.org/en/rsf-s-2023-round-45-journalists-killed-line-duty-worldwide-drop-despite-tragedy-gaza)][[3](https://unesdoc.unesco.org/ark:/48223/pf0000379589.locale=en)]</sup>. As an Information Security consultant, I've trained such users to use [TAILS](https://tails.net/) with Tor Browser -- so that all of their internet traffic is protected with [Tor](https://community.torproject.org/user-research/personas/). And I recommend that they use Tor Browser in "Safest" mode, which disables javascript for additional protection.

Using Tor Browser is the best way for at-risk users to stay safe when using the Internet

Unfortunately it appears that Twenty Twenty-Four has a bug where at-risk users following security best-practices cannot navigate Twenty Twenty-Four websites because the menu doesn't work.

#### Solution

Navigation is neither a bell nor a whistle. Being able to open a menu isn't something that should work only for users with javascript enabled.

For components as critical as basic site navigation, we should provide a <noscript> fallback that opens the menu with CSS.

 * https://en.wikipedia.org/wiki/Progressive_enhancement 

#### Screenshots

Please see the following screenshots taken of the official [Twenty Twenty-Four demo site](https://wp-themes.com/twentytwentyfour/), taken on 2025-02-27 in Tor Browser v14.0.6 (based on Firefox v128.7.0esr) in Debian 12.

| [![Screenshot of Twenty Twenty-Four website in Tor Browser in "Standard" mode](https://github.com/user-attachments/assets/aca56a6d-9147-4944-a138-6edc24d17cc4)](https://github.com/user-attachments/assets/aca56a6d-9147-4944-a138-6edc24d17cc4) | [![Screenshot of Twenty Twenty-Four website in Tor Browser in "Safest" mode](https://github.com/user-attachments/assets/a8385099-c0f3-425c-b190-6c8b58398a40)](https://github.com/user-attachments/assets/a8385099-c0f3-425c-b190-6c8b58398a40) |
|:--:|:--:|
| In Tor Browser "Standard" Mode (with javascript enabled), the menu works | In Tor Browser "Strict" Mode (with javascript disabled), the menu does not work |

**Comment by carolinan** [comment]
Discussions about how many people are assassinated are really unnecessary and may be triggering,
The repositories need to be family friendly as they are public **and also available to all ages.**

**Comment by carolinan** [comment]
@maltfield I believe a noscript element would need to go in here https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/navigation/index.php

**Comment by afercia** [comment]
To me, a no-JS menu with maybe some reduced functionality would be a great addition.

While not 100% semantically appropriate, maybe a `details` element could be used within a `noscript` to reveal a menu with plain links given the details works even with JS disabled. Demo: https://codepen.io/afercia/full/WbNjqBP

**Comment by stokesman** [comment]
#63917 demonstrates a way the menu can work without JS.

**Comment by westonruter** [comment]
Yes, I think per https://github.com/WordPress/gutenberg/pull/63917#issuecomment-2759102442 we have an opportunity to leverage [Invoker Commands](https://developer.mozilla.org/en-US/docs/Web/API/Invoker_Commands_API) to declaratively open a nav menu implemented in a `dialog` element. Browser support is not quite here yet, but assuming it is implemented in Chromium soon-ish it will also be easy to polyfill (with JavaScript). So while it wouldn't work without JavaScript for everyone in the immediate term, in the medium to long term it would work and would be a fully functional implementation that wouldn't depend on JavaScript and it wouldn't offer a degraded experience.

**Comment by westonruter** [comment]
Related: Core-50182 (Majority of bundled themes have broken mobile nav menus when JS turned off)

**Comment by maltfield** [comment]
@westonruter do you know if hardened browsers in strict mode (eg Tor Browser) will disable Invoker Commands, just as they disable javascript -- for security? I don't understand why we'd consider implementing this simple feature in some new technology.

Why not use CSS menus as a fallback (or just replace the JS entirely)? There's tons of CSS menus available online that we can just copy & paste. It works in almost all browsers today, and it's very future-proof to work in all browsers decades from now. And I don't know of any security-hardened browsers that have disabled CSS..

**Comment by maltfield** [comment]
Here's an example of a CSS menu:

 * https://codepen.io/kirstenhumphreys/pen/vgaKmG

**Comment by westonruter** [comment]
@maltfield Good questions. As I understand, Invoker Commands are not scripting so they would still work when scripting is disabled. That's really the whole point of them. Invoker Commands are declarative and are not "Turing complete" so there aren't the security concerns like with running arbitrary JavaScript. That said, Invoker Commands are currently in the process of being implemented, so we'll need to look out for this. I see that Tor Browser is based on Firefox, and Firefox currently has Invoker Commands implemented in a Nightly build if you have the `dom.element.invokers.enabled` flag enabled. I hope to test that this week.

I don't think a pure CSS mobile nav menu will be suitable for accessibility reasons. In that [Codepen](https://codepen.io/kirstenhumphreys/pen/vgaKmG), there's no setting of the required ARIA attributes which should be [handled automatically](https://open-ui.org/components/future-invokers.explainer/#accessibility) by Invoker Commands connecting a `button` with a `dialog`/`popover`. I don't think the current UI for the nav menu would be feasible to implement in pure CSS nav menu and it wouldn't be able to do things like focus trapping or light dismiss (e.g. hit <kbd>ESC</kbd> to close the nav menu).

**Comment by carolinan** [comment]
Why does it need to be the same UI if JavaScript is not enabled? Should it not be enough that you can access the content?

**Comment by westonruter** [comment]
Possibly, although I'm not a designer. Also, as I said I don't think a pure CSS nav menu is ideal for accessibility. I think we're on the cusp of being at a point on the web platform where we can have a fully accessible nav menu that has complete visual and functional parity with the current JS-based nav menu, but which doesn't require JavaScript at all.

**Comment by westonruter** [comment]
Here you can see with the Dialog element and Invoker Commands there is complete visual and functional parity when JavaScript is turned off: https://github.com/WordPress/gutenberg/pull/69720#issuecomment-2770475295

**Comment by carolinan** [comment]
Well its not an option until it is actually supported.

**Comment by westonruter** [comment]
The good news is that supported now in Chrome and is implemented in Firefox and Safari TP behind flags ([caniuse](https://caniuse.com/mdn-html_elements_button_commandfor)). So I expect it to be available in all the major browsers sometime in the next quarter. By the time the Navigation block implementation is finished (https://github.com/WordPress/gutenberg/pull/69720), Invoker Commands will surely be available across all the major browsers according to what [WordPress supports](https://make.wordpress.org/core/handbook/best-practices/browser-support/) by the time of the next WordPress major release, especially if there is a change to the release cadence.

**Comment by maltfield** [comment]
@westonruter before you commit to a new & uncertain technology, can you please [reach out to the Tor Project](https://forum.torproject.org/) and ask them if they plan on blocking Invoker Commands in Tor Browser or not?

This ticket was, after all, [created for providing at-risk users](https://github.com/WordPress/gutenberg/issues/69363#issuecomment-2691464563) a way to navigate wodpress sites. If Tor Browser blocks it (as they do with many features from upstream firefox), then this ticket will not be resolved.

**Comment by westonruter** [comment]
@maltfield I created a Tor forum topic and I've gotten very positive responses: https://forum.torproject.org/t/support-for-the-upcoming-invoker-commands-api-in-tor-browser-when-js-is-disabled/18208

**Comment by maltfield** [comment]
> The good news is that supported now in Chrome and is implemented in Firefox and Safari TP behind flags (caniuse). So I expect it to be available in all the major browsers sometime in the next quarter. 

@westonruter two quarters have past, and [caniuse](https://caniuse.com/mdn-html_elements_button_commandfor) still says it's not enabled for Firefox. So then it's also probably not available on Tor Browser (which is a modified version of Firefox).

Do you have a link to the relevant Firefox ticket so we can check to see when this will become enabled on vanilla firefox?

Edit: I think it's [this one](https://bugzilla.mozilla.org/show_bug.cgi?id=1856430)? Not many comments there. Not sure what's the hold-up?

**Comment by westonruter** [comment]
@maltfield coincidentally, this Intent to Ship just popped up: https://groups.google.com/a/mozilla.org/g/dev-platform/c/ck7taWZO2cs/m/MbukFk9gCwAJ

This is the bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1856430

So it looks like we're very close!

**Comment by westonruter** [comment]
And Invoker Commands are now landing in Safari! https://webkit.org/blog/17640/webkit-features-for-safari-26-2/

Also supported in Firefox since version 144.

Current global client support is over 70%: https://caniuse.com/mdn-html_elements_button_commandfor
